### PR TITLE
fix: error when spec code registers nested hooks

### LINKED ISSUE #8214: Cypress 4.10 breaks nested hooks - Cannot read property 'addCommand' of undefined
Steps:

* Create a support/testus.js file with this content

```js
module.exports = function something() {

  beforeEach(() => {
    cy.log('hier');
  });
};
```

* Create a support/index.js file with this content

```js
const something = require('./testus.js');

beforeEach(() => {
  cy.log('hier');
  something();
});

```

* Create a simple test like 

```js
describe('A', function() {
  it.only('B', function () {
    assert.fail('foo');
  });
});

```

* Run test with Cypress 4.9 IDE

Expected, as in Cypress 4.9 and previous versions it should log 2 messages.

```
BEFORE EACH

log hier
log testus
...
```

image

* Run test with Cypress 4.10 and above (I also tried the pre release sha 7341476f0f3c2c548969d1e9cbff08038c41e7a9)

```
BEFORE EACH
log hier
TypeError
Cannot read property 'addCommand' of undefined

Because this error occurred during a before each hook we are skipping all of the remaining tests.
```

image

Stack trace:

    cypress_runner.js:197883 TypeError: Cannot read property 'addCommand' of undefined
    
    Because this error occurred during a `before each` hook we are skipping all of the remaining tests.
        at Test.addCommand (https://somepage.com/__cypress/runner/cypress_runner.js:183869:12)
        at https://somepage.com/__cypress/runner/cypress_runner.js:183368:27
        at RunnablesStore._withTest (https://somepage.com/__cypress/runner/cypress_runner.js:183408:17)
        at RunnablesStore.addLog (https://somepage.com/__cypress/runner/cypress_runner.js:183367:18)
        at EventEmitter.<anonymous> (https://somepage.com/__cypress/runner/cypress_runner.js:181702:22)
        at executeAction (https://somepage.com/__cypress/runner/cypress_runner.js:49450:13962)
        at EventEmitter.n (https://somepage.com/__cypress/runner/cypress_runner.js:49450:13826)
        at EventEmitter.emit (https://somepage.com/__cypress/runner/cypress_runner.js:51593:5)
        at $Cypress.<anonymous> (https://somepage.com/__cypress/runner/cypress_runner.js:197692:19)
        at $Cypress.../driver/node_modules/eventemitter2/lib/eventemitter2.js.EventEmitter.emit (https://somepage.com/__cypress/runner/cypress_runner.js:84157:19)
        at $Cypress.parent.<computed> [as emit] (https://somepage.com/__cypress/runner/cypress_runner.js:170331:33)
        at $Cypress.action (https://somepage.com/__cypress/runner/cypress_runner.js:163930:28)
        at trigger (https://somepage.com/__cypress/runner/cypress_runner.js:171343:22)
        at triggerLog (https://somepage.com/__cypress/runner/cypress_runner.js:171349:12)
        at $Cypress.logFn [as log] (https://somepage.com/__cypress/runner/cypress_runner.js:171418:5)
        at logFn (https://somepage.com/__cypress/runner/cypress_runner.js:163715:26)
    From previous event:
        at run (https://somepage.com/__cypress/runner/cypress_runner.js:167240:16)
        at $Cy.cy.<computed> [as log] (https://somepage.com/__cypress/runner/cypress_runner.js:167666:11)
        at __stackReplacementMarker (https://somepage.com/__cypress/runner/cypress_runner.js:166756:13)
        at Context.runnable.fn (https://somepage.com/__cypress/runner/cypress_runner.js:167891:21)
        at callFn (https://somepage.com/__cypress/runner/cypress_runner.js:103248:21)
        at Hook.../driver/node_modules/mocha/lib/runnable.js.Runnable.run (https://somepage.com/__cypress/runner/cypress_runner.js:103235:7)
        at https://somepage.com/__cypress/runner/cypress_runner.js:173106:28
    From previous event:
        at Object.onRunnableRun (https://somepage.com/__cypress/runner/cypress_runner.js:173094:20)
        at $Cypress.action (https://somepage.com/__cypress/runner/cypress_runner.js:163887:61)
        at Hook.Runnable.run (https://somepage.com/__cypress/runner/cypress_runner.js:171707:13)
        at next (https://somepage.com/__cypress/runner/cypress_runner.js:103750:10)
        at https://somepage.com/__cypress/runner/cypress_runner.js:103789:7
        at next (https://somepage.com/__cypress/runner/cypress_runner.js:173030:16)
        at https://somepage.com/__cypress/runner/cypress_runner.js:173051:11
    From previous event:
        at onNext (https://somepage.com/__cypress/runner/cypress_runner.js:173048:57)
        at done (https://somepage.com/__cypress/runner/cypress_runner.js:103188:5)
        at https://somepage.com/__cypress/runner/cypress_runner.js:103253:11
    From previous event:
        at callFn (https://somepage.com/__cypress/runner/cypress_runner.js:103251:14)
        at Hook.../driver/node_modules/mocha/lib/runnable.js.Runnable.run (https://somepage.com/__cypress/runner/cypress_runner.js:103235:7)
        at https://somepage.com/__cypress/runner/cypress_runner.js:173106:28
    From previous event:
        at Object.onRunnableRun (https://somepage.com/__cypress/runner/cypress_runner.js:173094:20)
        at $Cypress.action (https://somepage.com/__cypress/runner/cypress_runner.js:163887:61)
        at Hook.Runnable.run (https://somepage.com/__cypress/runner/cypress_runner.js:171707:13)
        at next (https://somepage.com/__cypress/runner/cypress_runner.js:103750:10)
        at https://somepage.com/__cypress/runner/cypress_runner.js:103794:5
        at timeslice (https://somepage.com/__cypress/runner/cypress_runner.js:97720:27)
    logError @ cypress_runner.js:197883
    (anonymous) @ cypress_runner.js:197530
    emit @ cypress_runner.js:51593
    (anonymous) @ cypress_runner.js:181785
    emit @ cypress_runner.js:51593
    emit @ cypress_runner.js:181828
    onPrint @ cypress_runner.js:180649
    _onPrintClick @ cypress_runner.js:180654
    (anonymous) @ cypress_runner.js:182039
    executeAction @ cypress_runner.js:49450
    n @ cypress_runner.js:49450
    ca @ cypress_runner.js:59116
    ja @ cypress_runner.js:59117
    ka @ cypress_runner.js:59117
    wa @ cypress_runner.js:59119
    Aa @ cypress_runner.js:59120
    ya @ cypress_runner.js:59120
    Da @ cypress_runner.js:59123
    Ad @ cypress_runner.js:59186
    Gi @ cypress_runner.js:59352
    Kb @ cypress_runner.js:59141
    Dd @ cypress_runner.js:59188
    (anonymous) @ cypress_runner.js:59353
    ../../node_modules/scheduler/cjs/scheduler.production.min.js.exports.unstable_runWithPriority @ cypress_runner.js:63493
    Ii @ cypress_runner.js:59353
    Cd @ cypress_runner.js:59187
    With Cypress 4.10 these 2 things are listed as features, and I think are the reason for the breakage. ```You can open a before, beforeEach, after, and afterEach hook definition in your IDE from the Test Runner's Command Log by clicking the Open in IDE button. Addresses #7793.before, beforeEach, after, and afterEach hook definitions now display separately in the Test Runner's Command Log when defined in separate hook definitions. Addresses #7779.```I do realize that it is an anti-pattern to have hooks like this defined instead of in the tests, but as long as I don't care in what order they are called i should be able to use them. Why do I have code like this? Well actually I went trough some of the plugins you have listed in your documentation and they do it. I borrowed the code to implement similar features specific to our product, which now stopped working and I am stuck with Cypress 4.9.0.This ticket is similar to https://github.com/cypress-io/cypress/issues/8086. The fix for https://github.com/cypress-io/cypress/issues/8086 is in https://github.com/cypress-io/cypress/pull/8113. As mentioned above I tried pre-release https://github.com/cypress-io/cypress/commit/7341476f0f3c2c548969d1e9cbff08038c41e7a9, which should include the fix from PR https://github.com/cypress-io/cypress/pull/8113, but it doesn't fix my use case above. This is what I used to test with the pre-release and as said, it still fails. ```export CYPRESS_INSTALL_BINARY=https://cdn.cypress.io/beta/binary/5.0.0/darwin-x64/circle-develop-7341476f0f3c2c548969d1e9cbff08038c41e7a9-414411/cypress.zip
npm install https://cdn.cypress.io/beta/npm/5.0.0/circle-develop-7341476f0f3c2c548969d1e9cbff08038c41e7a9-414404/cypress.tgz`

### LINKED ISSUE #8288: Cypress server StatusCodeError: 412 on hookId and hookName is required
### Current behavior:
When we run cypress in parallel, `cypress@4.10.0` and above receives service error during `Uploading Results`. 
[Cypress dashboard](https://dashboard.cypress.io/) will be pending and never finish. 

See the below error messages:

```
  (Uploading Results)

We encountered an unexpected error talking to our servers.

Because you passed the --parallel flag, this run cannot proceed because it requires a valid response from our servers.

The --ciBuildId flag you passed was: xxx

The server's response was:

StatusCodeError: 412

{
  "message": "Request should follow putInstanceRequest@2.1.0 schema",
  "errors": [
    "data.hooks.1.hookId is required",
    "data.hooks.1.hookName is required"
  ],
  "object": {
    "stats": {
      "suites": 8,
      "tests": 16,
      "passes": 16,
      "pending": 0,
      "skipped": 0,
      "failures": 0,
      "wallClockStartedAt": "2020-08-14T20:33:38.948Z",
      "wallClockEndedAt": "2020-08-14T20:34:28.554Z",
      "wallClockDuration": 49606
    },
    "tests": [
      {
        "testId": "r4",
        "title": [
          "Generic Landing Page",
          "Landing Page Components",
          "Should be present on load"
        ],
        "state": "passed",
```

### Desired behavior:
Tested different cypress versions includes `4.9.0`, `4.10.0` and `4.12.1`. `4.9.0` works fine without issues.
Looking at the change log https://github.com/cypress-io/cypress/releases/tag/v4.10.0 
It includes hooks updates. That might cause the server error issue.
We have downgrade cypress from the latest version to `4.9.0` to run tests with cypress server.

### Test code to reproduce
```
cypress run --headless --browser chrome --record --key=<key> --ci-build-id=<build_id> --parallel --config baseUrl=<url>
```

### Versions

`4.10.0` and above

### LINKED ISSUE #8350: Nesting hooks within other hooks causes the test to never finish (runs forever)
### Current behavior:
Since upgrading to Cypress 5.0.0, I cannot get any tests to pass. Tests will run through all the code, but when they get to the end, the test does not get marked as passed and the timer keeps counting up.

image

### Desired behavior:
Tests end and move on the next test/spec.

### Test code to reproduce
```js
context('Never-ending Test', function () {
  it('Test', function () {
    cy.visit('http://wikipedia.com')
    cy.get('#searchInput')
      .type('Cypress.io')
    cy.get('.pure-button')
      .click()
    // Test never ends.
  })

  it('Next Test', function () {
    // Test never starts.
    cy.visit('http://wikipedia.com')
  })
})
```

### Versions
Cypress 5.0.0
macOS: 10.14.6
Browser: Chrome 84.0.4147.125, Firefox 79, Electron 83

┆Issue is synchronized with this [Jira Features](https://cypress-io.atlassian.net/browse/TR-282) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by kuceb**
we now throw an error message when we detect a nested hook (or hook registered after a test has already begun execution)
- Close #8214 
- Close #8288 
- Address #8350
### User facing changelog
Bug Fix: we now throw a proper error message when we detect hooks that are registered while a test is running

### Additional details
- nested hooks is a subset of the real issue - registering a hook after a test has begun executing

#### Before

**obscure error**

Screen Shot 2020-08-26 at 3 17 41 PM

**run forever**

Screen Shot 2020-08-26 at 3 18 03 PM

#### After

**errors msg**

Screen Shot 2020-08-26 at 3 18 56 PM

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub?

### CODE REVIEW

File: packages/driver/src/cypress/runner.js [hunk 1]
```diff
@@ -674,6 +675,14 @@ const _runnerListeners = (_runner, Cypress, _emissions, getTestById, getTest, se
...
 
   _runner.on('hook', (hook) => {
+    if (hook.hookId == null) {
+      hook.hookId = getHookId()
+    }
+
+    if (hook.hookName == null) {
+      hook.hookName = getHookName(hook)
+    }
+
```
Comment (Bkucera) [inline packages/driver/src/cypress/runner.js:685]: no longer needed

**Review by jennifer-shehane — COMMENTED** [review]
Manually verified this is working. I think the error message is a bit unclear on what exactly is wrong, but I get that it's a fine line between technically accurate and actionable.

**Comment by jennifer-shehane** [comment]
@Bkucera The test is failing legitimately with the new error you added.

File: packages/driver/src/cypress/cy.js [hunk 2]
```diff
@@ -713,7 +713,8 @@ const create = function (specWindow, Cypress, Cookies, state, config, log) {
     // attached to the error thrown from chai
     // command errors and command assertion errors (default assertion or cy.should)
     // have the invocation stack attached to the current command
-    let userInvocationStack = state('currentAssertionUserInvocationStack')
+    // prefer err.userInvocation stack if it's been set
+    let userInvocationStack = $errUtils.getUserInvocationStack(err) || state('currentAssertionUserInvocationStack')
```
Comment (chrisbreiding) [inline packages/driver/src/cypress/cy.js:717]: Why `$errUtils.getUserInvocationStack(err)` and not just `err.userInvocationStack`?

File: packages/driver/src/cypress/stack_utils.js [hunk 3]
```diff
@@ -288,7 +296,7 @@ const normalizedUserInvocationStack = (userInvocationStack) => {
   // whereas Chromium browsers have the user's line first
   const stackLines = getStackLines(userInvocationStack)
   const winnowedStackLines = _.reject(stackLines, (line) => {
-    return line.includes('cy[name]') || line.includes('Chainer.prototype[key]')
+    return line.includes('cypress:///')
```
Comment (chrisbreiding) [inline packages/driver/src/cypress/stack_utils.js:299]: I'm not sure about this change. Keep in mind `cypress:///` only appears in development and not production, so while the tests may pass for this, the effect will be different in production.

File: packages/runner/cypress/integration/issues/issue-8350.js [hunk 4]
```diff
@@ -1,20 +1,21 @@
...
+describe('issue-8350', { viewportHeight: 900 }, () => {
+  const file = 'nested_hooks_spec.js'
 
-      describe('s1', () => {
-        it('t1', () => {
-          //
-        })
-      })
-    })
+  verify.it('test', {
```
Comment (chrisbreiding) [inline packages/runner/cypress/integration/issues/issue-8350.js:10]: Can we use a better test title that describes what it's testing? Like 'shows a proper error message when a beforeEach is nested in a before'?

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: packages/driver/src/cypress/stack_utils.js [hunk 5]
```diff
@@ -111,6 +111,14 @@ const getCodeFrameFromSource = (sourceCode, { line, column, relativeFile, absolu
   }
 }
 
+const captureUserInvocationStack = (Error, userInvocationStack) => {
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cypress/stack_utils.js:114]: good point

File: packages/driver/src/cypress/cy.js (hunk 2 above)
Comment (Bkucera) [reply packages/driver/src/cypress/cy.js:717]: brian's reasoning was we dont want magical properties on the error instance without going through an explicit contract

File: packages/driver/src/cypress/stack_utils.js (hunk 3 above)
Comment (Bkucera) [reply packages/driver/src/cypress/stack_utils.js:299]: ok, well then this wont work...

File: packages/driver/src/cypress/stack_utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (brian-mann) [reply packages/driver/src/cypress/stack_utils.js:114]: agreed ^^

File: packages/driver/src/cypress/stack_utils.js [hunk 6]
```diff
@@ -308,4 +316,157 @@ module.exports = {
   stackWithoutMessage,
   stackWithReplacementMarkerLineRemoved,
   stackWithUserInvocationStackSpliced,
+  captureUserInvocationStack,
 }
+
+/**
```
[the remark this answers is not in the document]
Comment (brian-mann) [reply packages/driver/src/cypress/stack_utils.js:322]: well this info was lost to an issue - and this is such essential/critical work i think it should go directly in the spot that deals with this. it took us a *long* time to work this out and i think its important to co-locate with the primary file responsible for this. its really the *why* we do everything

**Review by brian-mann — CHANGES_REQUESTED** [review]
Please add a reference line to `errorScenarios.md` in various files that utilize that logic - like a `See: ./errorScenarios.md for details`

Otherwise it'll be completely opaque and unfindable.

**Review by brian-mann — DISMISSED** [review]

**Review by chrisbreiding — APPROVED** [review]

**Pull request merged**